// @datadog/browser-rum-core/esm/domain/rumEventsCollection/view/viewCollection.js
import { getTimestamp, isEmptyObject, mapValues, msToNs } from '@datadog/browser-core';
import { RumEventType } from '../../../rawRumEvent.types';
import { LifeCycleEventType } from '../../lifeCycle';
import { trackViews } from './trackViews';
export function startViewCollection(lifeCycle, location, onNewLocation) {
    lifeCycle.subscribe(LifeCycleEventType.VIEW_UPDATED, function (view) {
        return lifeCycle.notify(LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, processViewUpdate(view));
    });
    return trackViews(location, lifeCycle, onNewLocation);
}
function processViewUpdate(view) {
    var viewEvent = {
        _dd: {
            document_version: view.documentVersion,
        },
        date: getTimestamp(view.startTime),
        type: RumEventType.VIEW,
        view: {
            action: {
                count: view.eventCounts.userActionCount,
            },
            cumulative_layout_shift: view.cumulativeLayoutShift,
            dom_complete: msToNs(view.timings.domComplete),
            dom_content_loaded: msToNs(view.timings.domContentLoaded),
            dom_interactive: msToNs(view.timings.domInteractive),
            error: {
                count: view.eventCounts.errorCount,
            },
            first_contentful_paint: msToNs(view.timings.firstContentfulPaint),
            first_input_delay: msToNs(view.timings.firstInputDelay),
            first_input_time: msToNs(view.timings.firstInputTime),
            is_active: view.isActive,
            name: view.name,
            largest_contentful_paint: msToNs(view.timings.largestContentfulPaint),
            load_event: msToNs(view.timings.loadEvent),
            loading_time: msToNs(view.loadingTime),
            loading_type: view.loadingType,
            long_task: {
                count: view.eventCounts.longTaskCount,
            },
            resource: {
                count: view.eventCounts.resourceCount,
            },
            time_spent: msToNs(view.duration),
        },
    };
    if (!isEmptyObject(view.customTimings)) {
        viewEvent.view.custom_timings = mapValues(view.customTimings, msToNs);
    }
    return {
        rawRumEvent: viewEvent,
        startTime: view.startTime,
    };
}
