// @datadog/browser-rum-core/cjs/domain/rumEventsCollection/view/viewCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startViewCollection = void 0;
var browser_core_1 = require("@datadog/browser-core");
var rawRumEvent_types_1 = require("../../../rawRumEvent.types");
var lifeCycle_1 = require("../../lifeCycle");
var trackViews_1 = require("./trackViews");
function startViewCollection(lifeCycle, location, onNewLocation) {
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.VIEW_UPDATED, function (view) {
        return lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, processViewUpdate(view));
    });
    return trackViews_1.trackViews(location, lifeCycle, onNewLocation);
}
exports.startViewCollection = startViewCollection;
function processViewUpdate(view) {
    var viewEvent = {
        _dd: {
            document_version: view.documentVersion,
        },
        date: browser_core_1.getTimestamp(view.startTime),
        type: rawRumEvent_types_1.RumEventType.VIEW,
        view: {
            action: {
                count: view.eventCounts.userActionCount,
            },
            cumulative_layout_shift: view.cumulativeLayoutShift,
            dom_complete: browser_core_1.msToNs(view.timings.domComplete),
            dom_content_loaded: browser_core_1.msToNs(view.timings.domContentLoaded),
            dom_interactive: browser_core_1.msToNs(view.timings.domInteractive),
            error: {
                count: view.eventCounts.errorCount,
            },
            first_contentful_paint: browser_core_1.msToNs(view.timings.firstContentfulPaint),
            first_input_delay: browser_core_1.msToNs(view.timings.firstInputDelay),
            first_input_time: browser_core_1.msToNs(view.timings.firstInputTime),
            is_active: view.isActive,
            name: view.name,
            largest_contentful_paint: browser_core_1.msToNs(view.timings.largestContentfulPaint),
            load_event: browser_core_1.msToNs(view.timings.loadEvent),
            loading_time: browser_core_1.msToNs(view.loadingTime),
            loading_type: view.loadingType,
            long_task: {
                count: view.eventCounts.longTaskCount,
            },
            resource: {
                count: view.eventCounts.resourceCount,
            },
            time_spent: browser_core_1.msToNs(view.duration),
        },
    };
    if (!browser_core_1.isEmptyObject(view.customTimings)) {
        viewEvent.view.custom_timings = browser_core_1.mapValues(view.customTimings, browser_core_1.msToNs);
    }
    return {
        rawRumEvent: viewEvent,
        startTime: view.startTime,
    };
}
