// @datadog/browser-rum-core/cjs/domain/rumEventsCollection/longTask/longTaskCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startLongTaskCollection = void 0;
var browser_core_1 = require("@datadog/browser-core");
var rawRumEvent_types_1 = require("../../../rawRumEvent.types");
var lifeCycle_1 = require("../../lifeCycle");
function startLongTaskCollection(lifeCycle) {
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.PERFORMANCE_ENTRY_COLLECTED, function (entry) {
        if (entry.entryType !== 'longtask') {
            return;
        }
        var rawRumEvent = {
            date: browser_core_1.getTimestamp(entry.startTime),
            long_task: {
                duration: browser_core_1.msToNs(entry.duration),
            },
            type: rawRumEvent_types_1.RumEventType.LONG_TASK,
        };
        lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, {
            rawRumEvent: rawRumEvent,
            startTime: entry.startTime,
        });
    });
}
exports.startLongTaskCollection = startLongTaskCollection;
